### PR TITLE
Feature/plugin configs

### PR DESCRIPTION
**Pull request by jakeberesford-palmetto**
## Pull request checklist

Before submitting your PR, please review the following checklist:

- [x] Consider adding a unit test if your PR resolves an issue.
- [x] All new and existing tests pass locally (`palm test`)
- [x] Lint (`palm lint`) has passed locally and any fixes were made for failures
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Breaking changes

- [ ] Check if this pull request introduces a breaking change

## What does this implement/fix? Explain your changes.

Adds the concept of configuration to plugins, this has become a necessary tool for plugins like palm-dbt where we want to enable configuration of certain features. See: https://github.com/palmetto/palm-dbt/pull/63

While it is entirely possible to implement configs for each plugin, it makes a lot more sense to provide an optional plugin config mechanism for configuration in palm. This will ensure consistency in implementations, and make the process of using those configs much nicer.

### How does it work?

- If a plugin specifies a config, palm will automatically load that config _at run time_.
- If the config does not exist, palm will call the abstract `set` method specified by the plugin to guide users through configuration. Implementors will need to define the set method for their plugin config.
- Once a config exists, it is made available to commands via `environment.plugin_config('plugin_name')`
- Plugin configs are validated using Pydantic models, these are defined by the plugin implementor. Invalid plugin configs will raise an error.

## Does this close any currently open issues?

Unblocks the changes in https://github.com/palmetto/palm-dbt/pull/63 - though a re-write of that config will be necessary.

## Any other comments?

I still need to write docs around this, here's a quick code example of what a plugin config might look like in the wild:

```python
class DbtConfigModel(BaseModel):
    local_artifacts: str
    prod_artifacts: str


class DbtPluginConfig(BasePluginConfig):
    def __init__(self):
        super().__init__("dbt", DbtConfigModel)

    def set(self) -> dict:
        config = {
            'local_artifacts': click.prompt(
                "Where are your local artifacts?", type=str
            ),
            'prod_artifacts': click.prompt("Where are your prod artifacts?", type=str),
        }
        self.write(config)
        return config


dbtPlugin = BasePlugin(
    name="dbt", command_dir=Path(__file__).parent / "commands", config=DbtPluginConfig()
)
```

## Where has this been tested?

**Operating System:** macOs

**Platform:** python3.9

### CODE REVIEW

**Comment by jakeberesford-palmetto** [comment]
One thing this PR does not address is how we want to call the config setter outside of the initial initialization. Basically a UI for updating configs... should be relatively simple, but the semantics of that command need a little thought.

File: docs/source/write-a-plugin.rst [hunk 1]
```diff
@@ -30,3 +30,59 @@ your plugin's root directory:
...
+
+  # Define the config object
+  class MyPluginConfig(BasePluginConfig):
+      def __init__(self):
+          super().__init__('my_plugin', MyPluginConfigModel)
+
+      def set(self):
+          return {
+              "my_config_value": click.prompt("Please enter a value for my_config_value", type=str)
+          }
```
Comment (jakeberesford-palmetto) [inline docs/source/write-a-plugin.rst:67]: One internal debate I'm having is whether it's better to require `MyPluginConfigModel` as an argument in the initializer, or to define an `@abstractproperty` for `model` instead?

Curious whether you all prefer what's above, or this?

```python
class CoreConfig(BasePluginConfig):
    def __init__(self):
        super().__init__("core")

    @property
    def model(self) -> BaseModel:
        class ConfigModel(BaseModel):
            local_artifacts: str
            prod_artifacts: str

        return ConfigModel

    def set(self) -> dict:
        click.echo("set core plugin config")
        config = {
            'local_artifacts': click.prompt(
                "Where are your local artifacts?", type=str
            ),
            'prod_artifacts': click.prompt("Where are your prod artifacts?", type=str),
        }
        self.write(config)
        return config
```

**Review by evandegler-palmetto — COMMENTED** [review]
This looks good to me. I understand most of what's happening here, but not all of it.

I'm happy to approve it, based on what I do understand -- I trust the work that you do and that you've thoroughly tested everything here. But I understand too if you want @mariahjrogers's eyes on this PR, as I know she'll understand more of what's happening here than I do.

File: docs/source/write-a-plugin.rst [hunk 2]
```diff
@@ -30,3 +30,59 @@ your plugin's root directory:
 
 **Note**: Due to the way plugins are used by palm, you will need to re-install
 the plugin every time you want to test changes to the plugin
+
+Creating a Plugin Config
+========================
+
+If you wantto make your plugin configurable, you will need to create a plugin
```
[the remark this answers is not in the document]
Comment (mariahjrogers) [reply docs/source/write-a-plugin.rst]: +1

File: docs/source/write-a-plugin.rst (hunk 1 above)
Comment (mariahjrogers) [reply docs/source/write-a-plugin.rst:67]: Yep I think I like current implementation better.

File: docs/source/write-a-plugin.rst [hunk 3]
```diff
@@ -30,3 +30,59 @@ your plugin's root directory:
...
+
+Here is an example of how to use the plugin config values in a plugin command:
+
+.. code:: python
+
+  @click.command()
+  @click.pass_obj
+  def my_command(environment):
+      plugin_config = env.plugin_config('my_plugin')
+      my_config_value = plugin_config.get('my_config_value', '')
```
Comment (mariahjrogers) [inline docs/source/write-a-plugin.rst]: Would like an example using the dot notation!

**Review by mariahjrogers — APPROVED** [review]
Woooo! Much excitement 😁

**Pull request merged**